Refactor/navbar

osmaczko opened the pull request:
### What does the PR do

fixes: https://github.com/status-im/status-desktop/issues/7167
fixes: https://github.com/status-im/status-desktop/issues/8463

- removed imperative filtering
- removed imperative size calculation
- removed all ugly hacks

### Affected areas

application navbar

### StatusQ checklist

- ~~[ ] add documentation if necessary (new component, new feature)~~
- [x] update sandbox app
  - in case of new component, add new component page
  - in case of new features, add variation to existing component page
  - nice to have: add it to the demo application as well
- ~~[ ] test changes in both light and dark theme?~~

### Screenshot of functionality (including design for comparison)

https://user-images.githubusercontent.com/33099791/204269090-704ced01-07f9-4293-9a75-b6a50ff05cbd.mp4

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | a5f9872e | [#1](https://ci.status.im/job/status-desktop/job/branches/job/imports/job/PR-8477/1/) | 2022-11-28 11:45:19 | ~3 min | `imports` | [:page_facing_up:`log`](https://ci.status.im/job/status-desktop/job/branches/job/imports/job/PR-8477/1/consoleText) |

jrainville commented on ui/StatusQ/src/StatusQ/Layout/StatusAppNavBar.qml:
```diff
@@ -1,195 +1,129 @@
+    implicitWidth: 78
+    implicitHeight: layout.implicitHeight
+
+    Rectangle {
```
If we need a rectangle, why not keep it as the root?

osmaczko replied:
Good point :+1:

jrainville commented on ui/StatusQ/src/StatusQ/Layout/StatusAppNavBar.qml:
```diff
@@ -1,195 +1,129 @@
+            Layout.alignment: Qt.AlignHCenter
+            color: Theme.palette.directColor7
+
+            visible: chatItemsListView.count && mmunityItemsListView.contentHeight > communityItemsListView.height
```
Massive typo or broken line? `mmunityItemsListView`

osmaczko replied:
Good catch! Massive type :eyes: Have no clue how it worked :thinking:

jrainville commented on ui/StatusQ/src/StatusQ/Layout/StatusAppNavBar.qml:
```diff
@@ -1,195 +1,129 @@
 
-        delegate: delegateChooser
-    }
+        Rectangle {
```
Can we give this guy an id or a comment to say it's a separator?

osmaczko replied:
I left it without id because 1) it is not referenced anywhere 2) there are two separators and I wanted to avoid names like `separator1` and `separator2`

Added ids to them anyway cause I came up with better names :D

jrainville commented on ui/StatusQ/src/StatusQ/Layout/StatusAppNavBar.qml:
```diff
@@ -1,195 +1,129 @@
-        anchors.topMargin: 48
-        anchors.bottom: navBarProfileButtonSlot.top
-        anchors.bottomMargin: navBarButtonSpacing
+        Item {
```
What does this Item do?

osmaczko replied:
This is a common practice to separate items in layouts, you can think of it as an invisible spacer.
When you resize the layout it behaves as below:

![image](https://user-images.githubusercontent.com/33099791/204353388-df9576e2-f488-42ba-981d-f65f22fad826.png)

jrainville commented on ui/StatusQ/src/StatusQ/Layout/StatusAppNavBar.qml:
```diff
@@ -1,195 +1,129 @@
 
-        onAboutToUpdateFilteredModel: {
-            statusAppNavBar.aboutToUpdateFilteredRegularModel()
+        Rectangle {
```
Can we give this guy an id or a comment to say it's a separator?

jrainville reviewed (commented):
Looks good. Mostly nit comments.

One important point, did you test the duplication bug using a release build. Only release builds showed the issue. I can test myself later.

osmaczko commented:
> I just tested and this PR also fixes the duplicated communities tada I'll close mine

@jrainville  Thanks for testing this PR in the release version :pray: I have also addressed your comments, please take a look.

jrainville commented on ui/StatusQ/src/StatusQ/Layout/StatusAppNavBar.qml:
```diff
@@ -1,195 +1,130 @@
-        anchors.bottomMargin: visible ? 12 : 0
-    }
+    QtObject {
+        id: d
```
I think we decided that the `d` Object should be put at the top of the root, but I could be mistaken. Not gonna ask for changes, just asking/pointing it out.

jrainville reviewed (approved):
Good job. Way cleaner

caybro reviewed (approved):
Looks good, nice cleanup!

osmaczko commented:
> 1. Currently `StatusAppNavBar` is part of `StatusQ` despite it's application-specific component implementing specific behaviour/layout of our client. It's against the idea of `StatusQ`. The motivation to keep it here is probably the fact that it's used in the demo app. The problem is that the demo app doesn't suits well to `StatusQ` too. At least if it's using app-specific components like nav bar.

I agree, having `StatusAppNavBar` with assumptions about the order of sections (chats, communities, general, profile placeholder) is not the best fit for `StatusQ`. It is used there in both the Sandbox app to display sections and by the Demo App to imitate the Status navbar. The question here is whether it was the idea of `StatusQ`? I guess rather not, since it would be limiting to how others build their own clients.

> 2. There is a single model with data for `StatusAppNavBar` coming from backend. The model unnecessarily combine things of different kinds. On UI side in `AppMain` we do the reverse thing - split back into logically separate models via filtering. Imo backend should expose complete data in possibly simplest form, without any assumptions about the usage on the UI side. Otherwise we can easily end up with situation like here.

Hm.. How should the backend expose the data otherwise? By three separate models, that is: ChatsSectionModel, CommunitiesSectionModel, OthersSectionModel? IMO current approach does not make any assumptions about the UI usage, it is up to qml how to split/sort/filter the data. I think "section" is a good abstraction, you only provide what is logically needed for certain UI components.

> 3. `StatusAppNavBar` probably could mange it's delegates. Even now that component is very specific. It's first role is not reusability but decomposition. So moving delegates into that component won't break that (non-existing) reusability anyway. But we could make AppMain a bit shorter.

I don't think moving delegates into components will work as Sandbox and Demo App both use different delegates.

> I don't say that we should do it in that PR. I wanted to discuss it a bit to be on the same page and plan further refactor of that part for the future. Just please share what do you think about those points.

Initially, I wanted to create `AppNavBar` with all the specifics hidden inside, like model split and delegates, but it would break Sandbox and Demo App. 

Having app-specific components in StatusQ slows us down significantly because we can't freely rework components. The idea of moving app-specific components out of StatusQ was discussed many, many times. It still comes back like a boomerang. I think we should cover truly reusable and non app-specific components by a certain section in the Storybook then remove Sandbox and Demo App altogether and then move app-specific StatusQ components back to status-desktop. 

The issues you have raised are valid but the scope of that PR was not to change how things work logically, but rather to fix broken QML component with bad patterns.
